File: 22FALL/451.sort-characters-by-frequency.py
#
# @lc app=leetcode id=451 lang=python3
#
# [451] Sort Characters By Frequency
#

# @lc code=start
from collections import Counter


class Solution:
    def frequencySort(self, s: str) -> str:
        return ''.join(x * y for x, y in sorted(Counter(s).items(), key=lambda x: (x[1], x[0]), reverse=True))
# @lc code=end

File: VSCode_CPP/451.sort-characters-by-frequency.cpp
/*
 * @lc app=leetcode id=451 lang=cpp
 *
 * [451] Sort Characters By Frequency
 */

// @lc code=start
class Solution {
public:
    string frequencySort(string s) {
        unordered_map<char, int> mp;
        for (auto x:s) {
            mp[x]++;
        }
        s = "";
        vector<pair<char,int>>vec(mp.begin(), mp.end());
        sort(vec.begin(),vec.end(), [](const pair<char,int> &a, const pair<char,int> &b) {
            return a.second > b.second;
        });
        for (auto x:vec) {
            int cnt = 0;
            while (cnt < x.second) {
                s += x.first;
                cnt++;
            }
        }
        return s;
    }
};
// @lc code=end

